### PR TITLE
Add Optimizely A/B Testing framework

### LINKED ISSUE #3988: Add Optimizely A/B Testing
Add Optimizely to the project and hook it up to Tracks

### PR DESCRIPTION
**Pull request by kwonye**
Fixes #3988 

To test:

1. First update your `gradle.properties` to contain `wp.optimizely_token`
2. When you get to the LoginActivity when building off this branch, you will see "Add self-hosted site" become "Not on WordPress.com?"

screenshot_20160425-154158

Needs review: @maxme and @daniloercoli (for the Tracks events)

### CODE REVIEW

**Comment by maxme** [comment]
> When you get to the LoginActivity when building off this branch, you will see "Add self-hosted site" become "Not on WordPress.com?"

I don't see that in code, is this something pulled from optimizely?

* Optimizely adds okio as a dependency, we must add an exception to the `lint.xml` file (should fix travis builds):
    ```
    <issue id="InvalidPackage">
        <ignore regexp="okio-1.6.0.jar" />
    </issue>
    ```

**Comment by kwonye** [comment]
> > When you get to the LoginActivity when building off this branch, you will see "Add self-hosted site" become "Not on WordPress.com?"

> I don't see that in code, is this something pulled from optimizely?

Yes, it's their "WSYWIG" string modification. You can see it in the Optimizely dashboard. I didn't want to add a random variable just to prove integration worked.

**Comment by kwonye** [comment]
> Optimizely adds okio as a dependency, we must add an exception to the lint.xml file (should fix travis builds)

Thanks for that, @maxme. Worked great at fixing the lint warnings, but Travis seems to be getting suck. Any tips?

**Comment by maxme** [comment]
> Yes, it's their "WSYWIG" string modification. You can see it in the Optimizely dashboard. I didn't want to add a random variable just to prove integration worked.

Sorry, I don't get it, I thought you disabled the "WYSIWYG" stuff in https://github.com/wordpress-mobile/WordPress-Android/pull/4016/commits/deb9a75441464707d006639e8050072bbff7d105 

Also I don't understand why you're initializing Optimizely in `WPMainActivity`? Should we init it on startup, so we can use it for sign in / sign up?

File: WordPress/src/main/java/org/wordpress/android/util/WPOptimizelyEventListener.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+    }
+
+    @Override
+    public void onOptimizelyExperimentViewed(OptimizelyExperimentData optimizelyExperimentData) {
+        Map<String, OptimizelyExperimentData> visitedExperiments = Optimizely.getVisitedExperiments();
+
+        for (String experimentId : visitedExperiments.keySet()) {
+            OptimizelyExperimentData experiment = visitedExperiments.get(experimentId);
+            HashMap<String, String> abTestProperties = new HashMap<>();
```
Comment (daniloercoli) [inline WordPress/src/main/java/org/wordpress/android/util/WPOptimizelyEventListener.java]: Seems that properties are not included when bump Analytics.

File: libs/analytics/WordPressAnalytics/src/main/java/org/wordpress/android/analytics/AnalyticsTrackerNosara.java [hunk 2]
```diff
@@ -495,6 +495,9 @@ public void track(AnalyticsTracker.Stat stat, Map<String, ?> properties) {
             case SITE_SETTINGS_DELETE_SITE_RESPONSE_ERROR:
                 eventName = "site_settings_delete_site_response_error";
                 break;
+            case ABTEST_START:
+                eventName = "abtest_start";
```
Comment (daniloercoli) [inline libs/analytics/WordPressAnalytics/src/main/java/org/wordpress/android/analytics/AnalyticsTrackerNosara.java]: Shouldn't this be named `abtest_started` or `ab_test_started`. 
cc @astralbodies

Comment (kwonye) [inline libs/analytics/WordPressAnalytics/src/main/java/org/wordpress/android/analytics/AnalyticsTrackerNosara.java]: According to the docs (in fg), it is `{platform}_abtest_start`

Comment (astralbodies) [inline libs/analytics/WordPressAnalytics/src/main/java/org/wordpress/android/analytics/AnalyticsTrackerNosara.java]: The Track library adds the prefix - but the question I think is the verb should indicate what happened so past-tense `abtest_started` makes more sense.

Comment (kwonye) [inline libs/analytics/WordPressAnalytics/src/main/java/org/wordpress/android/analytics/AnalyticsTrackerNosara.java]: That's fair. I decided on `start` for consistency between this and Calypso/PHP. I'll defer to your guys' judgement and change it to `started` cc @aerych

Comment (astralbodies) [inline libs/analytics/WordPressAnalytics/src/main/java/org/wordpress/android/analytics/AnalyticsTrackerNosara.java]: Oh if they're using start for their event verb, might as well stick with it. Jerkfaces!

File: WordPress/src/main/java/org/wordpress/android/util/WPOptimizelyEventListener.java (hunk 1 above)
Comment (kwonye) [inline WordPress/src/main/java/org/wordpress/android/util/WPOptimizelyEventListener.java]: Oh my lord, I can't believe I did that. Thanks. I'll add it.

File: libs/analytics/WordPressAnalytics/src/main/java/org/wordpress/android/analytics/AnalyticsTrackerNosara.java (hunk 2 above)
Comment (daniloercoli) [inline libs/analytics/WordPressAnalytics/src/main/java/org/wordpress/android/analytics/AnalyticsTrackerNosara.java]: I agree with all of you :) Let's use `start` for consistency with Calypso/PHP.

**Comment by kwonye** [comment]
@maxme because the Gravatar feature is not yet in `develop`, I've decided to just change a string for the first A/B test. Once it is merged, we can A/B test the "change profile picture" row.

To see the variations in the Optimizely dashboard, you have to add `Optimizely.enableEditor()` in the first line of `WPMainActivity.configureOptimizely()`

> Also I don't understand why you're initializing Optimizely in WPMainActivity? 

Optimizely recommends activating it in the last line of the Main Activity's `onCreate()`. In `WPMainActivity` we are activating it *right* before the logic for showing the login screen, so all users will see it.

> Should we init it on startup, so we can use it for sign in / sign up?

If you want to move it to `WPLaunchActivity.onCreate()`, I have no problem with that.

**Comment by maxme** [comment]
> If you want to move it to WPLaunchActivity.onCreate(), I have no problem with that.

Yes please. Because the launch of `WPLaunchActivity` before the login screen is a just a trick to avoid duplicate code, but it also makes the time to first screen a bit longer - something we probably want to change in the future.

**Comment by kwonye** [comment]
@maxme moved to WPLaunchActivity in 82b9b8a 👍 ⚡

**Comment by maxme** [comment]
Looks good, we just need to update the checksum in `release-checks.sh` https://github.com/wordpress-mobile/WordPress-Android/issues/4044

:shipit:

**Pull request merged**